rfcs/79-minimal-fragmenting-relationships.md

```diff
@@ -0,0 +1,126 @@
+##### Query and System Caches
+Query's caches of matching tables and archetypes are updated each time the query is accessed by iterating through the archetypes that have been created since it was last accessed. As the number of archetypes in bevy is expected to stabilize over the runtime of a program this isn't currently an issue. However with the increased archetype fragmentation caused by fragmenting relationships and the new wrinkle of archetype deletion the updating of the query caches is poised to become a performance bottleneck.
+
+Luckily observers [#10839](https://github.com/bevyengine/bevy/pull/10839) can help solve this problem. By sending ECS events each time an archetype is created and creating an observer for each query we can target the updates to query caches so that we no longer need to iterate every new archetype. Similarly we can fire archetype deletion events to cause queries to remove the archetype from their cache.
+
+In order to implement such a mechanism we need some way for the observer to have access to the query state in order to update it. This can be done by moving the query state into a component on the associated entity (WIP branch here: [queries-as-entities](https://github.com/james-j-obrien/bevy/tree/queries-as-entities)) and will require further refactors to the cache to optimize for removal. However this is largely blocked by the deletion of entities in the render world tracked here: [#12144](https://github.com/bevyengine/bevy/issues/12144).
+
+Similarly systems cache their `ArchetypeComponentId` accesses for the purposes of multithreading, we also need to update theses caches to remove deleted ids presenting many of the same problems. 
+##### Access Bitsets and Component SparseSets
+The `FixedBitSet` implementation used by `Access` inside of systems, system params, queries etc. relies on dense component ids to avoid using a large amount of memory. If we were to start inserting ids with arbitrarily large entity targets as well as component ids shifted into the upper bits each `FixedBitSet` would allocate non-trivial amounts of memory to have flags for all the intervening ids. In order to minimize the affects of this we could consider all `(R, *)` pairs as one id in accesses, reducing the amount we need to track at the expense of some theoretical ability to parallelise.
+
+Similarly the `SparseSet` implementation used in both table and sparse component storage also operate under the assumption that component ids will remain low and would allocate an even larger amount of memory storing indexes for all the intervening ids when storing relationship pairs on an entity.
+
+The most practical way to solve this is using paged versions of the appropriate data structures, for bitsets this can take the form of existing hierarchical or compressed bitsets like `roaring`.
```

Comment:
Paging only resolves this within a fixed scalar factor. If we use 64-element pages, we incur 64x less memory used, but if the IDs are using bit shifted values, that's only going to account for 6 bits out of the 32 we're using at the very minimum, resulting in *very* large SparseSets being allocated. 

More care to ensure that encoded IDs in the sparse set is required (i.e. using morton encoding), and some acknowlegement of the potential costs of computing this when doing these lookups is warranted.

Reply:
For clarification, I don't use sparse sets anywhere in Flecs where a pair can be used as key, as even with a paged sparse set that could allocate lots of pages because of how wide the range of possible values is.

I'm using hashmaps wherever pairs are used as keys, sometimes combined with an array for the lower id ranges where performance matters.

Reply:
That's true. For access bitsets it should be feasible when collapsing all `(R, *)` as we effectively only double our id space, we just have to deal with them being sparse.

Reply:
From what I understand the way lookups stay fast in `flecs` without needing the sparse lookup in the table is using the component index so I suppose I should also detail that here.